File: Wells/Source/Filters/BooleanFilter.vb
Imports Wells.CorePersistence.Repositories

Public Class BooleanFilter(Of T)
    Inherits BaseFilter(Of T)

    Property Value As Boolean

    Public Overrides ReadOnly Property DisplayValue As String
        Get
            Return If(Value, "Verdadero", "Falso")
        End Get
    End Property

    Public Overrides ReadOnly Property IsDateRangeFilter As Boolean
        Get
            Return False
        End Get
    End Property

    Overrides ReadOnly Property Description As String
        Get
            Return $"{DisplayPropertyName} = {DisplayValue}"
        End Get
    End Property

    Sub New(propertyName As String, displayName As String, repo As IBussinessObjectRepository, value As Boolean)
        MyBase.New(propertyName, displayName, repo)
        _Value = value
    End Sub

    Public Overrides Function Apply(originalList As IQueryable(Of T)) As IQueryable(Of T)
        Return From o In originalList
               Where CBool(CallByName(o, PropertyName, CallType.Get)) = _Value
               Select o
    End Function
End Class

File: Wells/src/Filters/BooleanFilter.cs
using Microsoft.VisualBasic;
using System.Collections.Generic;
using System.Linq;
using System.Xml;
using Wells.Base;
using Wells.View.ViewModels;

namespace Wells.View.Filters
{
    public class BooleanFilter<T> : BaseFilter<T>
    {
        public bool Value { get; set; }
        public override string DisplayValue => Value ? "Verdadero" : "Falso";

        public override bool IsDateRangeFilter => false;

        public override string Description
        {
            get
            {
                return $"{DisplayPropertyName} = {DisplayValue}";
            }
        }

        public BooleanFilter() { }

        public BooleanFilter(string propertyName, string displayName, bool value) :
            base(propertyName, displayName)
        {
            Value = value;
        }

        public override IEnumerable<T> Apply(IEnumerable<T> queryable)
        {
            return from o in queryable
                   where (bool)(Interaction.CallByName(o, PropertyName, CallType.Get)) == Value
                   select o;
        }

        public override void SetUpdatedValues(FilterViewModel filterViewModel)
        {
            Value = filterViewModel.BooleanValue;
        }

        public override void ReadXml(XmlReader reader)
        {
            base.ReadXml(reader);
            Value = reader.ReadElementContentAsBoolean();
            reader.ReadEndElement();
        }

        public override void WriteXml(XmlWriter writer)
        {
            base.WriteXml(writer);
            writer.Write(nameof(Value), Value);
        }
    }
}
